File: app/src/main/java/com/aexon/theme/foundation/AexonChrome.java
package com.aexon.theme.foundation;

import androidx.annotation.ColorInt;

import com.aexon.theme.hct.utils.AexonColorUtils;

public class AexonChrome {
	private final double hue;
	private final double chroma;
	
	private AexonChrome(double hue, double chroma) {
		this.hue = hue;
		this.chroma = chroma;
	}
	
	public double getHue() {
		return hue;
	}
	
	public double getChroma() {
		return chroma;
	}
	
	public static AexonChrome fromInt(@ColorInt int argb) {
		double[] lab = AexonColorUtils.labFromArgb(argb);
		double a = lab[1];
		double b = lab[2];
		
		double hue = Math.toDegrees(Math.atan2(b, a));
		if (hue < 0) hue += 360.0;
		
		double chroma = Math.sqrt(a * a + b * b);
		return new AexonChrome(hue, chroma);
	}
}

File: app/src/main/java/com/aexon/theme/foundation/AexonChrome.kt
package com.aexon.theme.foundation

import com.aexon.theme.hct.utils.AexonColorUtils
import kotlin.math.atan2
import kotlin.math.sqrt

class AexonChrome private constructor(val hue: Double, val chroma: Double) {
    companion object {
        fun fromInt(argb: Int): AexonChrome {
            val lab = AexonColorUtils.labFromArgb(argb)
            val a = lab[1]
            val b = lab[2]

            var hue = Math.toDegrees(atan2(b, a))
            if (hue < 0) hue += 360.0

            val chroma = sqrt(a * a + b * b)
            return AexonChrome(hue, chroma)
        }
    }
}